src/module.c

```diff
@@ -13489,6 +13492,96 @@ int RM_GetDbIdFromDefragCtx(RedisModuleDefragCtx *ctx) {
     return ctx->dbid;
 }
 
+/* --------------------------------------------------------------------------
+ * ## Key API for Set type
+ *
+ * See also RM_ValueLength(), which returns the length of a set.
+ * -------------------------------------------------------------------------- */
+
+/* Operate a set key.
+ * REDISMODULE_SET_ADD: Add new elements into a set. If the key is an empty key
```

Comment:
I think SetOperate with an extra parameter to select the operation, and that the return value depends on the operation, is a weird API. I prefer RM_SetAdd, etc. It's possible to use SetOperate as a helper if it simplifies the implementation. The implementation in general looks good to me.

This is just my opinion. Better wait for @oranagra's comment before you change it again. He is probably busy with the release of redis 7.2 now. This PR will be in redis 8 so there is no hurry.

Reply:
If we were starting from scratch, I don't know I would have a strong opinion about the API. But we already have explicit APIs (HashGet and HashSet for e.g.) for all of the other types, so it would be weird that this one API have a different format (SetOperation). So I guess that leads me to have a slight preference for having 3 different APIs with common code underneath, even though the vargs makes it verbose (but you can have the helper take in vargs).

Also coming at it from someone that writes a lot of rust (which, maybe ignore), I prefer explicit signatures and documentation for functions. I don't think I mind duplicating the flags, especially since there is just one. EBADF also only makes sense for SADD and SREM, so can be cleaned up. If you were writing code in something like rust which requires complete error handling, you wouldn't need to handle EBADF for ISMEMBER operation, which leads me to think rust binds for these APIs would want distinct functions with custom return codes for each type (or at least the read and write split), which leads me to believe we probably should have separate functions.